sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -92,52 +141,46 @@ class BeamFnControlServicer(beam_fn_api_pb2_grpc.BeamFnControlServicer):
   _DONE_MARKER = object()
 
   def __init__(self):
-    self._push_queue = queue.Queue()
-    self._futures_by_id = dict()
-    self._read_thread = threading.Thread(
-        name='beam_control_read', target=self._read)
+    self._lock = threading.Lock()
     self._uid_counter = 0
     self._state = self.UNSTARTED_STATE
-    self._lock = threading.Lock()
+    # following self._req_* variables are used for debugging purpose, data is
+    # added only when self._log_req is True.
+    self._req_sent = collections.defaultdict(int)
+    self._req_worker_mapping = {}
+    self._log_req = logging.getLogger().getEffectiveLevel() <= logging.DEBUG
+    self.connections_by_worker_id = collections.defaultdict(ControlConnection)
 
   def Control(self, iterator, context):
     with self._lock:
       if self._state == self.DONE_STATE:
         return
       else:
         self._state = self.STARTED_STATE
-    self._inputs = iterator
-    # Note: We only support one client for now.
-    self._read_thread.start()
+
+    worker_id = dict(context.invocation_metadata()).get('worker_id')
+    if not worker_id:
+      raise RuntimeError('All workers communicate through gRPC should have '
+                         'worker_id. Received None.')
+
+    control_conn = self.connections_by_worker_id[worker_id]
```

Comment:
I just realized that defaultdicts may not be threadsafe, which could cause a rare but hard-to-track-down bug here. 

We should make this private and move all direct references to this dict to a method that grabs a lock before accessing it. (Similarly for the data channel.)